pkg/grafana/folder-handler.go

```diff
@@ -87,6 +89,34 @@ func (h *FolderHandler) GetUID(resource grizzly.Resource) (string, error) {
 	return resource.Name(), nil
 }
 
+type folderForSorting struct {
+	UID      string
+	Resource grizzly.Resource
+}
+
+// Sort sorts according to handler needs
+func (h *FolderHandler) Sort(resources grizzly.Resources) grizzly.Resources {
+	foldersByUID := map[string]grizzly.Resource{}
+	graph := topsort.NewGraph()
+	rootFolders := []grizzly.Resource{}
+	for _, resource := range resources {
+		foldersByUID[resource.UID()] = resource
+		if parentUID, ok := resource.Spec()["parentUid"]; ok {
+			graph.AddEdge(parentUID.(string), resource.UID())
+		} else {
+			rootFolders = append(rootFolders, resource)
+		}
+	}
+	sorted := grizzly.Resources{}
+	for _, resource := range rootFolders {
+		s, _ := graph.TopSort(resource.UID())
+		for i := len(s) - 1; i >= 0; i-- {
+			sorted = append(sorted, foldersByUID[s[i]])
+		}
+	}
+	return sorted
+}
+
```

Comment:
```suggestion

// Sort sorts according to handler needs
func (h *FolderHandler) Sort(resources grizzly.Resources) grizzly.Resources {
	result := grizzly.Resources{}
	addedToResult := map[string]bool{}
	for _, resource := range resources {
		addedToResult[resource.UID()] = false
	}
	for {
		continueLoop := false
		for _, resource := range resources {
			parentUID, hasParentUID := resource.Spec()["parentUid"]
			// Add root folders
			if !hasParentUID {
				addedToResult[resource.UID()] = true
				result = append(result, resource)
				continue
			}
			parentAdded, parentExists := addedToResult[parentUID.(string)]
			// Add folders with parents which aren't declared in Grizzly
			if !parentExists {
				addedToResult[resource.UID()] = true
				result = append(result, resource)
				continue
			}
			// Add folders with parents which have already been added
			if parentAdded {
				addedToResult[resource.UID()] = true
				result = append(result, resource)
				continue
			}
			// Delay folders with parents which haven't been added yet
			continueLoop = true
		}

		if !continueLoop {
			break
		}
	}

	return result
}
```

I feel like you don't need an external lib with possibly undefined behavior for this (ex: missing node). And you'll never have enough resources that an algorithm like this will be slow. This likely takes under a second for millions of folders

Reply:
Or with less lines:
```suggestion
// Sort sorts according to handler needs
func (h *FolderHandler) Sort(resources grizzly.Resources) grizzly.Resources {
	result := grizzly.Resources{}
	addedToResult := map[string]bool{}
	for _, resource := range resources {
		addedToResult[resource.UID()] = false
	}
	for {
		continueLoop := false
		for _, resource := range resources {
			parentUID, hasParentUID := resource.Spec()["parentUid"]
			parentAdded, parentExists := addedToResult[parentUID.(string)]
			// Add
			// - root folders
			// - folders with parents which aren't declared in Grizzly
			// - folders with parents which have already been added
			if !hasParentUID || !parentExists || parentAdded {
				addedToResult[resource.UID()] = true
				result = append(result, resource)
				continue
			}
			// Delay folders with parents which haven't been added yet
			continueLoop = true
		}

		if !continueLoop {
			break
		}
	}

	return result
}
```

Reply:
Thx, will review.

Reply:
@julienduchesne I committed your proposal, with one addition: you need to check for values already added to the addedToResult map and skip them.